lucene/core/src/test/org/apache/lucene/search/TestTopFieldCollectorEarlyTermination.java

```diff
@@ -234,13 +242,189 @@ public void testCanEarlyTerminateOnPrefix() {
+        segmentSizes.remove(seg);
+        segDocsToFill.remove(seg);
+      } else {
+        ++seg;
+      }
+    }
+    final long seed = random().nextLong();
+    final IndexWriterConfig iwc = newIndexWriterConfig(new MockAnalyzer(new Random(seed)));
+    // one segment per commit so we can control the segment sizes
+    iwc.setMergePolicy(NoMergePolicy.INSTANCE);
+    iwc.setIndexSort(sort);
+    iw = new RandomIndexWriter(new Random(seed), dir, iwc);
+    for (int seg = 0; seg < segDocs.size(); seg++) {
+      for (Document doc : segDocs.get(seg)) {
+        iw.addDocument(doc);
+      }
+      iw.commit();
+    }
+    reader = iw.getReader();
+  }
+
+  private void createUniformIndex() throws IOException {
+    dir = newDirectory();
+    numDocs = atLeast(150);
+    int numSegs = atLeast(5);
+    // Create segments of random pre-determined sizes so we can distribute the documents uniformly
+    // among them
+    int docsRemaining = numDocs;
+    List<Integer> segmentSizes = new ArrayList<>();
+    for (int i = 0; i < numSegs - 1; i++) {
+      int size = random().nextInt(docsRemaining - numSegs + i);
+      segmentSizes.add(size);
+      docsRemaining -= size;
+    }
+    segmentSizes.add(docsRemaining);
+    createRandomTerms();
+    final long seed = random().nextLong();
+    final IndexWriterConfig iwc = newIndexWriterConfig(new MockAnalyzer(new Random(seed)));
+    // one segment per commit so we can control the segment sizes
+    iwc.setMergePolicy(NoMergePolicy.INSTANCE);
+    iwc.setMaxBufferedDocs(numDocs);
+    iwc.setRAMBufferSizeMB(IndexWriterConfig.DISABLE_AUTO_FLUSH);
+    iwc.setIndexSort(sort);
+    writer = new IndexWriter(dir, iwc);
+    for (int seg = 0; seg < numSegs; seg++) {
+      int size = segmentSizes.get(seg);
+      double step = numDocs / (double) size;
+      for (int i = 0; i < size; i++) {
+        int num = (int) Math.round(i * step);
+        Document doc = numberedDocument(num, num);
+        writer.addDocument(doc);
+      }
+      writer.commit();
+    }
+    reader = DirectoryReader.open(writer);
+  }  
+
+  private void createSkewedIndex() throws IOException {
+    dir = newDirectory();
+    numDocs = atLeast(150);
+    createRandomTerms();
+    final long seed = random().nextLong();
+    final IndexWriterConfig iwc = newIndexWriterConfig(new MockAnalyzer(new Random(seed)));
+    // one segment per commit so we can control the segment sizes
+    iwc.setMergePolicy(NoMergePolicy.INSTANCE);
+    iwc.setIndexSort(sort);
+    writer = new IndexWriter(dir, iwc);
+    for (int i = 0; i < numDocs; ++i) {
+      // insert the documents in order, so successive segments have increasingly larger documents
+      writer.addDocument(numberedDocument(i, i));
+      if (random().nextInt(numDocs / 10) == 0) {
+        // Make about 10 random-sized segments
+        writer.commit();
+      }
+    }
+    reader = DirectoryReader.open(writer);
+  }
+
+  public void testProratedEarlyTermination() throws IOException {
```

Comment:
Do we explicitly test passing `Integer.MAX_VALUE` for `perSegmentMargin` somewhere here?

Reply:
hmm, explicitly .. no. It's the default now, so tested implicitly in many cases, but an explicit test would be good to demonstrate all still works OK even when index distribution is skewed: I'll add that.